### PR TITLE
Add support for rescoring searches.

### LINKED ISSUE #311: How to rescore a search?
Is [rescoring](https://www.elastic.co/guide/en/elasticsearch/guide/current/_improving_performance.html#rescore-api) on searches supported? Something like:

```
Search().query("match", title="my title").rescore(window_size=50, ...)
```

### PR DESCRIPTION
**Pull request by fandrews**
Closes #311

### CODE REVIEW

**Comment by HonzaKral** [comment]
Thanks for this PR, do you think you could also add some documentation on this feature? A short example in the docs + more elaborate docstring with an example would be super useful.

**Comment by fxdgear** [comment]
@fandrews Would be nice if you could update this PR as there are some conflicts that need to be resolved.

**Comment by erip** [comment]
@HonzaKral What's the status of this? What kind of documentation is remaining? I'd love to see this merged.

File: elasticsearch_dsl/search.py [hunk 1]
```diff
@@ -587,6 +605,9 @@ def to_dict(self, count=False, **kwargs):
             if self._script_fields:
                 d['script_fields'] = self._script_fields
 
+            if self._rescore:
+                d['rescore'] = self._rescore
```
Comment (HonzaKral) [inline elasticsearch_dsl/search.py]: Can you also add complementary logic to `update_from_dict` (plus related test) here? Thanks!

File: elasticsearch_dsl/search.py [hunk 2]
```diff
@@ -526,6 +528,22 @@ def highlight(self, *fields, **kwargs):
             s._highlight[f] = kwargs
         return s
 
+    def rescore(self, query, **kwargs):
+        """
+        Add a rescore request to the search.
+
+        :arg query: query to be used during the rescore process.
```
Comment (HonzaKral) [inline elasticsearch_dsl/search.py]: It would be great to have a working example snippet here as documentation

**Comment by HonzaKral** [comment]
@erip thanks for the feedback (and sorry for the late reply), I think just a short example snippet in the docstring should be enough, then also adding the `update_from_dict` part...

**Comment by karmi** [comment]
Hi @fandrews, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/elasticsearch-dsl-py/pull/415.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

**Comment by fandrews** [comment]
Hi @karmi, a colleague @gregoryhooks assisted with the changes that were requested.  I have asked him to complete the license agreement.

**Comment by sethmlarson** [comment]
Hey! I'm going to close this PR as stale for now, the original author(s) or others are welcome to make changes here or submit a new PR as this is still a desirable feature.

**Pull request closed without merging**